blog: introducing kube-scheduler-simulator

sanposhiho opened the pull request:
This PR adds a blog post to introduce [sigs/kube-scheduler-simulator](https://github.com/kubernetes-sigs/kube-scheduler-simulator).

dipesh-rawat commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,113 @@
+title: "Introducing Kube-scheduler-simulator"
+date: 2024-07-xx
+slug: introducing-kube-scheduler-simulator
+canonicalUrl: https://kubernetes.dev/blog/2024/07/xx/introducing-kube-scheduler-simulator/
```
What's the reason for including the `canonicalURL` here? Is this blog mirrored from the [k8s.dev](https://www.kubernetes.dev/blog/) site? If so, review comments should be added to the original content there.

sanposhiho replied:
Sorry, actually I just copied those tags from another blog post. 

I saw Kwok's blog post, which I actually copied tags from, have two articles in two websites.
- https://www.kubernetes.dev/blog/2023/03/01/introducing-kwok/
- https://kubernetes.io/blog/2023/03/01/introducing-kwok/

Would you recommend creating a blog post in kubernetes.dev too? What's the general difference between those two websites?

sftim replied:
https://www.kubernetes.dev/blog/ is aimed only at people contributing to Kubernetes
https://www.kubernetes.io/blog/ is aimed at people using to Kubernetes

Who is this article aimed at? I think people might be using this even though they are not Kubernetes contributors - right?

sanposhiho replied:
> Who is this article aimed at? I think people might be using this even though they are not Kubernetes contributors - right?

Yes, correct. So, here's the place for kubernetes.io blog, right?
[Did you suggest](https://github.com/kubernetes/website/pull/47104#pullrequestreview-2214335131) making the same post on .dev too? Or, is having it only here fine?

sftim replied:
This article is in the right repo. No need to change it.

We never mirror from the general website to the contributor website, so there's also no need to open a PR against the contributor site for this article.

dipesh-rawat commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,113 @@
+The simulator allows users to create Kubernetes resources and observe how each plugin influences the scheduling decisions for Pods.
+This visibility helps users understand the scheduler’s workings and define appropriate scheduling constraints.
+
+![Screenshot of the simulator web frontend](./simulator.png)
```
_Improvement:_ To display the image incorporate a '[figure](https://gohugo.io/content-management/shortcodes/#figure)' shortcode with a caption and maybe also include an alt attribute for better accessibility. Take a look at how the shortcode is utilized in a [previous blog post (here)](https://github.com/kubernetes/website/blob/fa563d70564885ae829abf6a228ed4d082a77f75/content/en/blog/_posts/2021-09-29-data-duplication-in-data-heavy-k8s-env.md?plain=1#L108).

dipesh-rawat commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,113 @@
+
+A big shout out to all [the awesome contributors](https://github.com/kubernetes-sigs/kube-scheduler-simulator/graphs/contributors)!
+
+![contributors](./contributors.png)
```
Same for this image; consider using '[figure](https://gohugo.io/content-management/shortcodes/#figure)' shortcode. Refer to how this shortcode was used in a [previous blog post (here)](https://github.com/kubernetes/website/blob/fa563d70564885ae829abf6a228ed4d082a77f75/content/en/blog/_posts/2021-09-29-data-duplication-in-data-heavy-k8s-env.md?plain=1#L108).

sanposhiho commented:
updated to use `fiture`

Okabe-Junya commented:
FYI: tracking issue - #39776

nate-double-u commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,114 @@
+
+[Kube-scheduler-simulator](sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, initially developed by Kensei Nakada as part of [Google Summer of Code 2021](https://summerofcode.withgoogle.com/). 
+This tool allows users to examine the scheduler’s behavior and decisions closely, 
+making it useful for both casual users who employ scheduling constraints (e.g., PodAffinity) 
```
```suggestion
It is useful for casual users who employ scheduling constraints (e.g., PodAffinity) 
```

nate-double-u commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,114 @@
+Thus, anyone utilizing Kubernetes relies on the scheduler.
+
+[Kube-scheduler-simulator](sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, initially developed by Kensei Nakada as part of [Google Summer of Code 2021](https://summerofcode.withgoogle.com/). 
+This tool allows users to examine the scheduler’s behavior and decisions closely, 
```
```suggestion
This tool allows users to closely examine the scheduler’s behavior and decisions.
```

nate-double-u commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,114 @@
+composed of many plugins that each contribute to the scheduling decision-making process from their unique perspectives. 
+Understanding its behavior can be challenging due to the multitude of factors it considers. 
+Even if a Pod seems to be scheduled as expected in your simple test cluster, 
+it’s maybe because of a different reason from what you imagine, 
+and could result in unexpected scheduling results in a large production environment.
```
```suggestion
it may be for a different reason than what you imagine, 
which could result in unexpected scheduling results in a large production environment.
```

nate-double-u commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,114 @@
+
+## What’s Kube-scheduler-simulator? 
+
+The simulator’s core feature is its ability to expose the internal decisions of the scheduler. 
```
```suggestion
The simulator’s core feature is its ability to expose the scheduler's internal decisions. 
```

nate-double-u commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,114 @@
+
+## What are the use cases?
+
+1. **Cluser users**: Examine if your scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.
```
```suggestion
1. **Cluster users**: Examine if your scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.
```

nate-double-u commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
````diff
@@ -0,0 +1,114 @@
+```
+
+Users can also integrate their custom plugins or extenders into the debuggable scheduler and visualize their results. 
+Also, this debuggable scheduler can run standalone, e.g., on any Kubernetes cluster or in integration tests. 
+Those would be useful to custom plugin developers to test their plugins or examine their custom scheduler in a real cluster with better debuggability.
````
```suggestion
This debuggable scheduler can also run standalone, e.g., on any Kubernetes cluster or in integration tests. 
This would be useful to custom plugin developers who want to test their plugins or examine their custom scheduler in a real cluster with better debuggability.
```

nate-double-u reviewed (commented):
a copy-edit pass and some nits:

sanposhiho commented:
Addressed the points from @nate-double-u.

Ritikaa96 commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,116 @@
+Inside the simulator, instead of the vanilla scheduler, 
+it runs the debuggable scheduler, which outputs each scheduler plugin’s result from every extension point to Pod’s annotations. 
+Inside the simulator, a debuggable scheduler runs instead of the vanilla scheduler. 
+This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations. 
+Actually, the web frontend just shows the results reading from the annotations.
```
```suggestion
This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations 
and the web front end shows the results fetched from the annotations.
```

Ritikaa96 commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,116 @@
+The Kube-scheduler-simulator is developed by Kubernetes SIG Scheduling. Your feedback and contributions are welcome!
+
+Open issues or PRs at [Kube-scheduler-simulator repository](sigs.k8s.io/kube-scheduler-simulator)
+Join the conversation on the #sig-scheduling slack channel.
```
```suggestion
Join the conversation on the [#sig-scheduling](https://kubernetes.slack.com/messages/sig-scheduling) slack channel.
```

Ritikaa96 reviewed (changes requested):
nit(optional)

Ritikaa96 commented:
The overall content LGTM.

sanposhiho commented:
Fixed.

network-charles commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,116 @@
+{{< figure src="/images/blog/2024-07-xx-kube-scheduler-simulator/simulator.png" alt="Screenshot of the simulator web frontend" title="The simulator web frontend" >}}
+
+Inside the simulator, instead of the vanilla scheduler, 
+it runs the debuggable scheduler, which outputs each scheduler plugin’s result from every extension point to Pod’s annotations. 
+Inside the simulator, a debuggable scheduler runs instead of the vanilla scheduler. 
```
```suggestion
it runs the debuggable scheduler, which outputs each scheduler plugin’s result from every extension point to the Pod’s annotations. 
```
The sentence in Line 45 is redundant with Lines 43-44.

network-charles commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,116 @@
+
+{{< figure src="/images/blog/2024-07-xx-kube-scheduler-simulator/simulator.png" alt="Screenshot of the simulator web frontend" title="The simulator web frontend" >}}
+
+Inside the simulator, instead of the vanilla scheduler, 
+it runs the debuggable scheduler, which outputs each scheduler plugin’s result from every extension point to Pod’s annotations. 
```
```suggestion
```
The sentence is redundant with Lines 45-46.

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,116 @@
+The Kubernetes Scheduler is a crucial control plane component that determines which Node a Pod will run on. 
+Thus, anyone utilizing Kubernetes relies on the scheduler.
+
+[Kube-scheduler-simulator](sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, initially developed by Kensei Nakada as part of [Google Summer of Code 2021](https://summerofcode.withgoogle.com/). 
```
```suggestion
[Kube-scheduler-simulator](https://sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, initially developed by Kensei Nakada as part of [Google Summer of Code 2021](https://summerofcode.withgoogle.com/).
```

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,116 @@
+
+You can then access the simulator's web UI at http://localhost:3000.
+
+Visit [Kube-scheduler-simulator repository](sigs.k8s.io/kube-scheduler-simulator) for more details!
```
```suggestion
Visit the [kube-scheduler-simulator repository](https://sigs.k8s.io/kube-scheduler-simulator) for more details!
```

sftim reviewed (commented):
Thanks.

Would you be willing to propose this for https://k8s.dev/blog instead? See https://github.com/kubernetes/contributor-site

sanposhiho commented:
Addressed the suggestions from @sftim @network-charles.

@sftim I actually had a question on that topic -

> Would you recommend creating a blog post in kubernetes.dev too? What's the general difference between those two blog websites?
https://github.com/kubernetes/website/pull/47104#discussion_r1671402207

Do you mean, we should move it to kubernetes.dev (and do nothing with .io)? Or we should have it on both websites?
(I actually saw Kwok's post is on both websites.)

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -24,9 +24,20 @@ Even if a Pod seems to be scheduled as expected in your simple test cluster,
+Also, testing your scheduler is a complex challenge.
+There are countless patterns of operations executed within a real cluster, making it impractical to anticipate every scenario with a finite number of tests. 
+More often than not, bugs are discovered only when the scheduler is deployed in an actual cluster.
+Actually, many bugs are reported by users after we ship the release, even in the upstream kube-scheduler. 
```
```suggestion
Actually, many bugs are reported by users after we ship the release, and the bugs are found
even in the upstream kube-scheduler. 
```

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -24,9 +24,20 @@ Even if a Pod seems to be scheduled as expected in your simple test cluster,
+More often than not, bugs are discovered only when the scheduler is deployed in an actual cluster.
+Actually, many bugs are reported by users after we ship the release, even in the upstream kube-scheduler. 
+
+Having a development or sandbox environment for testing the scheduler — or, indeed, any Kubernetes controllers — is a common practice for this problem.
```
```suggestion
Having a development or sandbox environment for testing the scheduler — or, indeed, any Kubernetes controllers — is a common practice.
```

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -24,9 +24,20 @@ Even if a Pod seems to be scheduled as expected in your simple test cluster,
+
+Having a development or sandbox environment for testing the scheduler — or, indeed, any Kubernetes controllers — is a common practice for this problem.
+However, this approach falls short of capturing all the potential scenarios that might arise in a production cluster 
+because a development cluster is often much smaller, and never sees the exact same use or exhibits the same behavior as its production counterpart, 
+with notable differences in workload sizes and scaling dynamics.
```
```suggestion
because a development cluster is often much smaller with notable differences in workload sizes and scaling dynamics.
It never sees the exact same use or exhibits the same behavior as its production counterpart.
```

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -24,9 +24,20 @@ Even if a Pod seems to be scheduled as expected in your simple test cluster,
+The scheduler simulator aims to solve those problems.
+It enables users to test their scheduling constraints, scheduler configurations, 
 and custom plugins while checking every detailed part of scheduling decisions.
+It also allows us to create a simulated cluster environment, where you can test your scheduler with the same resources as your production cluster, but without affecting your actual workloads.
```
```suggestion
It also allows you to create a simulated cluster environment, where you can test your scheduler
with the same resources as your production cluster without affecting your actual workloads.
```

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -76,11 +87,20 @@ Users can also integrate their custom plugins or extenders into the debuggable s
 
+## The simulator as a better dev cluster
+
+As mentioned earlier, it is impossible to predict every possible scenario in a real-world cluster with a limited set of tests. Typically, we test the scheduler in a small, development cluster before deploying it to production, hoping no issues arise.
```
```suggestion
As mentioned earlier, with a limited set of tests, it is impossible to predict every possible scenario in a real-world cluster.
Typically, you will test the scheduler in a small, development cluster before deploying it to production, hoping that no issues arise.
```

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -76,11 +87,20 @@ Users can also integrate their custom plugins or extenders into the debuggable s
+
+As mentioned earlier, it is impossible to predict every possible scenario in a real-world cluster with a limited set of tests. Typically, we test the scheduler in a small, development cluster before deploying it to production, hoping no issues arise.
+
+[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md) provides a solution by allowing you to simulate deploying a new scheduler version in a production-like environment without impacting your live workloads.
```
```suggestion
[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md)
provides a solution by allowing you to simulate deploying a new scheduler version in a production-like environment without impacting your live workloads.
```

sftim commented:
See https://github.com/kubernetes/website/pull/47104#discussion_r1750471818

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
````diff
@@ -0,0 +1,137 @@
+and the web front end shows the results fetched from the annotations.
+
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  annotations:
+    scheduler-simulator/bind-result: '{"DefaultBinder":"success"}'
+    scheduler-simulator/filter-result: >-
+      {"node-jjfg5":{"NodeName":"passed","NodeResourcesFit":"passed","NodeUnschedulable":"passed","TaintToleration":"passed"},"node-mtb5x":{"NodeName":"passed","NodeResourcesFit":"passed","NodeUnschedulable":"passed","TaintToleration":"passed"}}
+    scheduler-simulator/finalscore-result: >-
+      {"node-jjfg5":{"ImageLocality":"0","NodeAffinity":"0","NodeResourcesBalancedAllocation":"52","NodeResourcesFit":"47","TaintToleration":"300","VolumeBinding":"0"},"node-mtb5x":{"ImageLocality":"0","NodeAffinity":"0","NodeResourcesBalancedAllocation":"76","NodeResourcesFit":"73","TaintToleration":"300","VolumeBinding":"0"}}
+    scheduler-simulator/permit-result: '{}'
+    scheduler-simulator/permit-result-timeout: '{}'
+    scheduler-simulator/postfilter-result: '{}'
+    scheduler-simulator/prebind-result: '{"VolumeBinding":"success"}'
+    scheduler-simulator/prefilter-result: '{}'
+    scheduler-simulator/prefilter-result-status: >-
+      {"AzureDiskLimits":"","EBSLimits":"","GCEPDLimits":"","InterPodAffinity":"","NodeAffinity":"","NodePorts":"","NodeResourcesFit":"success","NodeVolumeLimits":"","PodTopologySpread":"","VolumeBinding":"","VolumeRestrictions":"","VolumeZone":""}
+    scheduler-simulator/prescore-result: >-
+      {"InterPodAffinity":"","NodeAffinity":"success","NodeResourcesBalancedAllocation":"success","NodeResourcesFit":"success","PodTopologySpread":"","TaintToleration":"success"}
+    scheduler-simulator/reserve-result: '{"VolumeBinding":"success"}'
+    scheduler-simulator/result-history: >-
+      [{"scheduler-simulator/bind-result":"{\"DefaultBinder\":\"success\"}","scheduler-simulator/filter-result":"{\"node-jjfg5\":{\"NodeName\":\"passed\",\"NodeResourcesFit\":\"passed\",\"NodeUnschedulable\":\"passed\",\"TaintToleration\":\"passed\"},\"node-mtb5x\":{\"NodeName\":\"passed\",\"NodeResourcesFit\":\"passed\",\"NodeUnschedulable\":\"passed\",\"TaintToleration\":\"passed\"}}","schedule ...
+    scheduler-simulator/score-result: >-
+      {"node-jjfg5":{"ImageLocality":"0","NodeAffinity":"0","NodeResourcesBalancedAllocation":"52","NodeResourcesFit":"47","TaintToleration":"0","VolumeBinding":"0"},"node-mtb5x":{"ImageLocality":"0","NodeAffinity":"0","NodeResourcesBalancedAllocation":"76","NodeResourcesFit":"73","TaintToleration":"0","VolumeBinding":"0"}}
+    scheduler-simulator/selected-node: node-mtb5x
````
If using this manifest unchanged, please call out that using the (unregistered) top level domain `scheduler-simulator` is **not** a recommended practice for Kubernetes object annotations. It's code from part of the Kubernetes project but it's not following Kubernetes good practice.

If possible, change the actual code to use a more appropriate domain name (eg `scheduler-simulator.sigs.k8s.io` - which would need [registering](https://kubernetes.io/docs/reference/labels-annotations-taints/)) **before** publication. If not feasible to do that, please explain that the domain name in the key is a known bug.

I don't want to tacitly endorse poor practice with an official-looking article.

---

BTW, you can register annotations with a PR against k/website. We aim for this to be easy; all you need say is that the annotation is used by https://sigs.k8s.io/kube-scheduler-simulator

sanposhiho replied:
Thanks, created https://github.com/kubernetes-sigs/kube-scheduler-simulator/issues/386.
We're about to ship a new release, and we can include the fix for it. The fix should be very easy.

sftim replied:
Annotations now registered BTW

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+As mentioned earlier, with a limited set of tests, it is impossible to predict every possible scenario in a real-world cluster.
+Typically, you will test the scheduler in a small, development cluster before deploying it to production, hoping that no issues arise.
+
+[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md)
```
```suggestion
The simulator's [importing](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/simulator/v0.3.0/simulator/docs/import-cluster-resources.md)
feature
```

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+
+## What are the use cases?
+
+1. **Cluster users**: Examine if your scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.
```
```suggestion
1. **Cluster users**: Examine if your scheduling constraints (e.g., Pod affinity, Pop topology spread) work as intended.
```

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+## What are the use cases?
+
+1. **Cluster users**: Examine if your scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.
+2. **Cluster admins**: Assess how your cluster would behave with changes to the scheduler configuration.
+3. **Scheduler plugin developers**: Test your custom scheduler plugins or extenders, use the debuggable scheduler in integration tests or development clusters, and use [the syncing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md) for testing within a production-like environment. 
```
```suggestion
1. **Cluster admins**: Assess how your cluster would behave with changes to the scheduler configuration.
1. **Scheduler plugin developers**: Test your custom scheduler plugins or extenders, use the debuggable scheduler in integration tests or development clusters, or use the [syncing](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/simulator/v0.3.0/simulator/docs/import-cluster-resources.md) feature for testing within a production-like environment.
```

- The site renderer does the numbering for you
- We avoid linking to trunk-like branches (_main_, _master_ etc) as the link is unlikely to be long term stable.

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+
+## Getting involved 
+
+The Kube-scheduler-simulator is developed by Kubernetes SIG Scheduling. Your feedback and contributions are welcome!
```
(nit)
```suggestion
The scheduler simulator is developed by Kubernetes SIG Scheduling. Your feedback and contributions are welcome!
```

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+
+The Kube-scheduler-simulator is developed by Kubernetes SIG Scheduling. Your feedback and contributions are welcome!
+
+Open issues or PRs at [Kube-scheduler-simulator repository](sigs.k8s.io/kube-scheduler-simulator)
```
(nit)
```suggestion
Open issues or PRs at the [kube-scheduler-simulator repository](https://sigs.k8s.io/kube-scheduler-simulator).
```

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+
+A big shout out to all [the awesome contributors](https://github.com/kubernetes-sigs/kube-scheduler-simulator/graphs/contributors)!
+
+![contributors](./contributors.png)
```
This actually doesn't need a `figure` shortcode, but _should_ have a more descriptive `alt` attribute.

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
````diff
@@ -0,0 +1,137 @@
+make docker_up
+```
+
+You can then access the simulator's web UI at http://localhost:3000.
````
(nit)
```suggestion
You can then access the simulator's web UI at `http://localhost:3000`.
```

sftim reviewed (commented):
Thanks for the PR, and please accept our apologies about the long wait for a review.

I hope this feedback comes over as helpful.

sftim commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,137 @@
+The Kubernetes Scheduler is a crucial control plane component that determines which Node a Pod will run on. 
+Thus, anyone utilizing Kubernetes relies on the scheduler.
+
+[Kube-scheduler-simulator](https://sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, initially developed by Kensei Nakada as part of [Google Summer of Code 2021](https://summerofcode.withgoogle.com/).
```
As this is a blog article, you can write in the first person, and you are implicitly writing as you.

network-charles commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
````diff
@@ -0,0 +1,137 @@
+    scheduler-simulator/selected-node: node-mtb5x
+```
+
+Users can also integrate their custom plugins or extenders into the debuggable scheduler and visualize their results. 
````
```suggestion
Users can also integrate their custom plugins or extenders, into the debuggable scheduler and visualize their results. 
```

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,135 @@
+author: Kensei Nakada (Tetrate)
+---
+
+The Kubernetes Scheduler is a crucial control plane component that determines which Node a Pod will run on. 
```
As a convention, we use 'Node' when we refer to the Node resource defined by the API, and we use 'node' as a regular term when we refer to the machine where workloads are
scheduled and run.

This is a style recommendation for all docs on the site, though not a strict one for
logs. Please check the doc when proposing your next commit to this PR.

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,135 @@
+Also, testing your scheduler is a complex challenge.
+There are countless patterns of operations executed within a real cluster, making it impractical to anticipate every scenario with a finite number of tests. 
+More often than not, bugs are discovered only when the scheduler is deployed in an actual cluster.
+Actually, many bugs are reported by users after we ship the release, and the bugs are found
```
Incomplete sentence?

tengqm commented on content/en/blog/_posts/2024-07-xx-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,135 @@
+## Features of the Kube-scheduler-simulator
+
+The simulator’s core feature is its ability to expose the scheduler's internal decisions.
+The scheduler operates based on the [Scheduling Framework](/concepts/scheduling-eviction/scheduling-framework/), 
```
```suggestion
The scheduler operates based on the [scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/), 
```

sanposhiho commented:
@tengqm @sftim How can I get it to move forward? It's been a long time.

tengqm commented:
/lgtm

sftim reviewed (commented):
We'll need to assign a publication date for this.

Good to know it looks good to publish, but let's ensure the date is right. How about the 22nd of January @sanposhiho ?

/lgtm cancel
because of that concern about a date

/hold
until we have an agreed, future publication date

sanposhiho commented:
@sftim Yes, looks great. I updated the date on the tag.

sanposhiho commented:
@sftim I think we missed the date. What's your next suggestion for the publish date?

sftim commented:
Because it's now marked as a draft
/hold cancel

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+because a development cluster is often much smaller with notable differences in workload sizes and scaling dynamics.
+It never sees the exact same use or exhibits the same behavior as its production counterpart.
+
+The scheduler simulator aims to solve those problems.
```
it is a first time you refer to `kube-scheduler-simulator` as "The scheduler simulator". Parhaps you can introduce the mapping between these two in introduction.

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+It also allows you to create a simulated cluster environment, where you can test your scheduler
+with the same resources as your production cluster without affecting your actual workloads.
+
+## Features of the Kube-scheduler-simulator
```
```suggestion
## Features of the kube-scheduler-simulator
```

lower case here OR expand into the full name, than Kubernetes is upper-cased

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+
+## Features of the Kube-scheduler-simulator
+
+The simulator’s core feature is its ability to expose the scheduler's internal decisions.
```
Now you refer to the scheduler simulator as just "simulator". A bit consistency will help to make this article a bit more understandable

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+
+The simulator’s core feature is its ability to expose the scheduler's internal decisions.
+The scheduler operates based on the [scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/), 
+utilizing various plugins at different extension points; 
```
is semicolon the right punctuation here?

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+The simulator allows users to create Kubernetes resources and observe how each plugin influences the scheduling decisions for Pods.
+This visibility helps users understand the scheduler’s workings and define appropriate scheduling constraints.
+
+{{< figure src="/images/blog/2025-01-22-kube-scheduler-simulator/simulator.png" alt="Screenshot of the simulator web frontend" title="The simulator web frontend" >}}
```
I think alt may need a bit more explanation on what screenshot is showing. So people who cannot see it understand what it highlights. Something like "Screenshot of the simulator web frontend showing three stages of scheduling: filtering, blah-blah".

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+
+{{< figure src="/images/blog/2025-01-22-kube-scheduler-simulator/simulator.png" alt="Screenshot of the simulator web frontend" title="The simulator web frontend" >}}
+
+Inside the simulator, a debuggable scheduler runs instead of the vanilla scheduler. 
```
what is vanilla scheduler in this context? Does it mean that the decision will be different between those two? If you mean the same scheduler, but in debug mode - it may be a better way to describe what is hapenning. Otherwise it is very confusing to read. Why I would want some other scheduler decisions for my investigation?

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+---
+layout: blog
+title: "Introducing Kube-scheduler-simulator"
+date: 2030-01-22
```
pick a reasonable date. I cannot see 2030 in the preview:

<img width="680" alt="image" src="https://github.com/user-attachments/assets/7e046c7d-09f8-4ce5-ae1e-b16d4ab68c09" />


Or tell me how to look at the preview

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
````diff
@@ -0,0 +1,138 @@
+This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations 
+and the web front end shows the results fetched from the annotations.
+
+```yaml
````
Can you point on how this is rendered? Cannot find it - see previous command

SergeyKanzhelev replied:
Can you describe what this code signifies BEFORE the code snippet? Otherwise it is not clear why it is here and what to look at there

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+    kube-scheduler-simulator.sigs.k8s.io/reserve-result: '{"VolumeBinding":"success"}'
+    kube-scheduler-simulator.sigs.k8s.io/result-history: >-
+      [{"kube-scheduler-simulator.sigs.k8s.io/bind-result":"{\"DefaultBinder\":\"success\"}","kube-scheduler-simulator.sigs.k8s.io/filter-result":"{\"node-jjfg5\":{\"NodeName\":\"passed\",\"NodeResourcesFit\":\"passed\",\"NodeUnschedulable\":\"passed\",\"TaintToleration\":\"passed\"},\"node-mtb5x\":{\"NodeName\":\"passed\",\"NodeResourcesFit\":\"passed\",\"NodeUnschedulable\":\"passed\",\"TaintTo ...
+    kube-scheduler-simulator.sigs.k8s.io/score-result: >-
```
I would expect the rendered version to be sompletely unreadable. Perhaps you can put some artificial newlines here for readability

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
````diff
@@ -0,0 +1,138 @@
+    kube-scheduler-simulator.sigs.k8s.io/selected-node: node-mtb5x
+```
+
+Users can also integrate their custom plugins or extenders, into the debuggable scheduler and visualize their results. 
````
what is extenders? Can you link docs for plugins and extenders?

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+Users can also integrate their custom plugins or extenders, into the debuggable scheduler and visualize their results. 
+
+This debuggable scheduler can also run standalone, e.g., on any Kubernetes cluster or in integration tests. 
+This would be useful to custom plugin developers who want to test their plugins or examine their custom scheduler in a real cluster with better debuggability.
```
Do you have an example that you can link? Is it actually used this way in some OSS projects that can be linked here?

SergeyKanzhelev replied:
As a minumum, link some readme here

sanposhiho replied:
I don't think any OSS project uses it, my previous company uses it internally though.

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+## The simulator as a better dev cluster
+
+As mentioned earlier, with a limited set of tests, it is impossible to predict every possible scenario in a real-world cluster.
+Typically, you will test the scheduler in a small, development cluster before deploying it to production, hoping that no issues arise.
```
sorry, who deployes scheduler to the production cluster? Do you mean some custom scheduler?

sanposhiho replied:
Yes.

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+Typically, you will test the scheduler in a small, development cluster before deploying it to production, hoping that no issues arise.
+
+[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md)
+provides a solution by allowing you to simulate deploying a new scheduler version in a production-like environment without impacting your live workloads.
```
who "you"?

sanposhiho replied:
I meant anyone who's using this project. What's the better way?

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md)
+provides a solution by allowing you to simulate deploying a new scheduler version in a production-like environment without impacting your live workloads.
+
+By continuously syncing between your production cluster and the simulator, you can safely test a new scheduler version with the same resources your production cluster handles. 
```
Is it referring to the scheduler maintainers? Or customer scheduler developers? Clarify "your" here. Are you suggesting Kubernertes scheduler maintainers to import their test cluster as a test to Kubernetes codebase?

sanposhiho replied:
Same, I mean "you" equals whoever use this project. Whoever use this project can test their custom scheduler with their production cluster's resources.

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+
+## Getting started
+
+The simulator only requires Docker to be installed on your machine; a Kubernetes cluster is not necessary.
```
how will it access the cluster for the use cases listed above? This is very confusing for me, I do not know what to expect when I run this. If it doesn't need a cluster, what scheduler version will it even use? Which feature gates enabled? Is it compatible with [my cloud] scheduler?

sanposhiho replied:
> how will it access the cluster for the use cases listed above?

It internally runs kwok cluster which you can access.

Regarding other points, we don't have compatible matrics or the feature gate control yet, which we have to improve.

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+---
+layout: blog
```
Do you want to refer the article: https://www.cncf.io/blog/2022/08/24/writing-your-own-scheduler-with-kube-scheduler-simulator/? Not sure what is the relashinship here between articles

sanposhiho replied:
I didn't know that article, and I don't want to, honestly.

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+
+## Getting involved 
+
+The scheduler simulator is developed by Kubernetes SIG Scheduling. Your feedback and contributions are welcome!
```
```suggestion
The scheduler simulator is developed by [Kubernetes SIG Scheduling](https://github.com/kubernetes/community/blob/master/sig-scheduling/README.md#kube-scheduler-simulator). Your feedback and contributions are welcome!
```

SergeyKanzhelev commented on content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,138 @@
+Also, testing your scheduler is a complex challenge.
+There are countless patterns of operations executed within a real cluster, making it impractical to anticipate every scenario with a finite number of tests. 
+More often than not, bugs are discovered only when the scheduler is deployed in an actual cluster.
+Actually, many bugs are reported by users after we ship the release, and the bugs are found
```
yes, clarification would be useful. Otherwise it is not clear who you mean here and who finds bugs where

sanposhiho commented:
Sorry for the delay here, I was occupied on k/k side.
Addressed @SergeyKanzhelev 's suggestions. cc @sftim

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+author: Kensei Nakada (Tetrate)
+---
+
+The Kubernetes Scheduler is a crucial control plane component that determines which node a Pod will run on. 
```
```suggestion
The [Kubernetes Scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) is a crucial control plane component that determines which node a Pod will run on. 
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+Understanding its behavior can be challenging due to the multitude of factors it considers. 
+Even if a Pod seems to be scheduled as expected in a simple test cluster, 
+it may be coming from a different calculation than the expectation, 
+which could result in unexpected scheduling results in a large production environment.
```
> Even if a Pod seems to be scheduled as expected in a simple test cluster, 
it may be coming from a different calculation than the expectation, 
which could result in unexpected scheduling results in a large production environment.

Try to make it like:

> Even if a Pod appears to be scheduled correctly in a simple test cluster, it might have been scheduled based on different calculations than expected. This discrepancy could lead to unexpected scheduling outcomes when deployed in a large production environment.

It's more readable.

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+which could result in unexpected scheduling results in a large production environment.
+
+Also, testing a scheduler is a complex challenge.
+There are countless patterns of operations executed within a real cluster, making it impractical to anticipate every scenario with a finite number of tests. 
```
```suggestion
There are countless patterns of operations executed within a real cluster, making it unfeasible to anticipate every scenario with a finite number of tests. 
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+because a development cluster is often much smaller with notable differences in workload sizes and scaling dynamics.
+It never sees the exact same use or exhibits the same behavior as its production counterpart.
+
+kube-scheduler-simulator aims to solve those problems.
```
```suggestion
The kube-scheduler-simulator aims to solve those problems.
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+
+## Features of the kube-scheduler-simulator
+
+kube-scheduler-simulator’s core feature is its ability to expose the scheduler's internal decisions.
```
```suggestion
The kube-scheduler-simulator’s core feature is its ability to expose the scheduler's internal decisions.
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+
+kube-scheduler-simulator’s core feature is its ability to expose the scheduler's internal decisions.
+The scheduler operates based on the [scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/), 
+utilizing various plugins at different extension points,
```
```suggestion
using various plugins at different extension points,
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+{{< figure src="/images/blog/2025-01-22-kube-scheduler-simulator/simulator.png" alt="Screenshot of the simulator web frontend that shows the detailed scheduling results per node and per extension point" title="The simulator web frontend" >}}
+
+Inside the simulator, a debuggable scheduler runs instead of the vanilla scheduler. 
+This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations like the following Yaml shows
```
```suggestion
This debuggable scheduler outputs the results of each scheduler plugin at every extension point to the Pod’s annotations like the following manifest shows
```

sftim replied:
(@sanposhiho you can make that change, and on this one I do recommend it, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+
+Users can also integrate [their custom plugins](/docs/concepts/scheduling-eviction/scheduling-framework/) or [extenders](https://github.com/kubernetes/design-proposals-archive/blob/main/scheduling/scheduler_extender.md), into the debuggable scheduler and visualize their results. 
+
+This debuggable scheduler can also run standalone, e.g., on any Kubernetes cluster or in integration tests. 
```
The style guides suggest to avoid latin phrases like "e.g" or "i.e"

https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases
```suggestion
This debuggable scheduler can also run standalone, for example, on any Kubernetes cluster or in integration tests. 
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+
+## What are the use cases?
+
+1. **Cluster users**: Examine if scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.
```
The style guides suggest to avoid latin phrases like "e.g" or "i.e"

https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases

```suggestion
1. **Cluster users**: Examine if scheduling constraints (for example PodAffinity, PodTopologySpread) work as intended.
```

sftim replied:
(@sanposhiho you can make that change, but it's not a blocker for publication)

graz-dev reviewed (commented):
@sanposhiho nice piece!
Just added some suggestion to improve readability, some typos or grammar/word choice fix and a style guide violation. 
Fixing these it could be published by my side @sftim .

sftim commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+The [kube-scheduler-simulator](https://sigs.k8s.io/kube-scheduler-simulator) is a simulator for the Kubernetes scheduler, started as a [Google Summer of Code 2021](https://summerofcode.withgoogle.com/) project developed by me (Kensei Nakada) and later received a lot of contributions.
+This tool allows users to closely examine the scheduler’s behavior and decisions. 
+
+It is useful for casual users who employ scheduling constraints (e.g., [inter-Pod affinity](/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity))
```
(@sanposhiho you can make that change, but it's not a blocker for publication)

sftim commented on content/en/blog/_posts/2025-12-31-kube-scheduler-simulator/index.md:
```diff
@@ -0,0 +1,220 @@
+---
+layout: blog
+title: "Introducing kube-scheduler-simulator"
```
We should get an answer about which way you'd like to go on this one @sanposhiho. It seems important we're using the right name that end users would typically know the software by.

sftim reviewed (commented):
Let's merge this as a draft.

@sanposhiho you can unhold this and have it merge, or you can make changes and request a new review.

After it merges, someone from the blog team ought to be in touch about getting it published (we can send in a small PR that sets the publication date and removes `draft: true` from the front matter).

/hold
/lgtm
/approve

sanposhiho commented:
/unhold

sanposhiho commented:
I'll make separate PR for the latest suggestions

sftim commented:
See https://github.com/kubernetes/website/pull/50218#pullrequestreview-2717200182
